File: apps/api/src/participants/participants.service.ts
import {
  Injectable,
  NotFoundException,
  ForbiddenException,
  ConflictException,
} from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { InviteParticipantDto } from './dto/participant.dto';

@Injectable()
export class ParticipantsService {
  constructor(private readonly prisma: PrismaService) {}

  async invite(decisionId: string, dto: InviteParticipantDto, actorId: string) {
    // Check decision exists and actor is owner
    const decision = await this.prisma.decision.findUnique({
      where: { id: decisionId },
      include: { participants: true },
    });

    if (!decision) {
      throw new NotFoundException('Decision not found');
    }

    const actorParticipant = decision.participants.find(
      (p) => p.userId === actorId,
    );

    if (!actorParticipant || actorParticipant.role !== 'owner') {
      throw new ForbiddenException('Only the owner can invite participants');
    }

    // Find user by email
    const invitedUser = await this.prisma.user.findUnique({
      where: { email: dto.email },
    });

    if (!invitedUser) {
      throw new NotFoundException('User not found with this email');
    }

    // Check if already a participant
    const existing = decision.participants.find(
      (p) => p.userId === invitedUser.id,
    );

    if (existing) {
      throw new ConflictException('User is already a participant');
    }

    // Create participant
    await this.prisma.participant.create({
      data: {
        decisionId,
        userId: invitedUser.id,
        role: dto.role,
      },
    });

    // Write audit log
    await this.prisma.auditLog.create({
      data: {
        entityId: decisionId,
        actorId,
        action: 'PARTICIPANT_INVITED',
        metadata: { invitedEmail: dto.email },
      },
    });

    // Return updated participants list
    return this.prisma.participant.findMany({
      where: { decisionId },
      include: {
        user: {
          select: { id: true, name: true, email: true },
        },
      },
    });
  }

  async remove(decisionId: string, targetUserId: string, actorId: string) {
    // Check decision exists and actor is owner
    const decision = await this.prisma.decision.findUnique({
      where: { id: decisionId },
      include: { participants: true },
    });

    if (!decision) {
      throw new NotFoundException('Decision not found');
    }

    const actorParticipant = decision.participants.find(
      (p) => p.userId === actorId,
    );

    if (!actorParticipant || actorParticipant.role !== 'owner') {
      throw new ForbiddenException('Only the owner can remove participants');
    }

    // Cannot remove yourself (the owner)
    if (targetUserId === actorId) {
      throw new ForbiddenException('Cannot remove yourself as the owner');
    }

    // Check target is a participant
    const targetParticipant = decision.participants.find(
      (p) => p.userId === targetUserId,
    );

    if (!targetParticipant) {
      throw new NotFoundException('Participant not found');
    }

    // Delete participant
    await this.prisma.participant.delete({
      where: {
        decisionId_userId: {
          decisionId,
          userId: targetUserId,
        },
      },
    });

    // Write audit log
    await this.prisma.auditLog.create({
      data: {
        entityId: decisionId,
        actorId,
        action: 'PARTICIPANT_REMOVED',
        metadata: { removedUserId: targetUserId },
      },
    });

    // Return updated participants list
    return this.prisma.participant.findMany({
      where: { decisionId },
      include: {
        user: {
          select: { id: true, name: true, email: true },
        },
      },
    });
  }
}

File: apps/api/app/services/participants_service.py
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.future import select
from sqlalchemy.orm import selectinload
from app.models.models import Participant, Decision, User, AuditLog
from app.schemas.schemas import InviteParticipantRequest
from fastapi import HTTPException, status
import uuid

async def invite_participant(db: AsyncSession, decision_id: str, dto: InviteParticipantRequest, current_user: User):
    try:
        did = uuid.UUID(decision_id)
    except ValueError:
        raise HTTPException(status_code=404, detail="Decision not found")

    result = await db.execute(select(Decision).filter(Decision.id == did))
    decision = result.scalars().first()
    if not decision:
        raise HTTPException(status_code=404, detail="Decision not found")

    owner_check = await db.execute(
        select(Participant).filter(Participant.decision_id == did, Participant.user_id == current_user.id, Participant.role == "owner")
    )
    if not owner_check.scalars().first():
        raise HTTPException(status_code=403, detail="Only owner can invite")

    user_result = await db.execute(select(User).filter(User.email == dto.email))
    target_user = user_result.scalars().first()
    if not target_user:
        raise HTTPException(status_code=404, detail="User not found")

    existing_p = await db.execute(
        select(Participant).filter(Participant.decision_id == did, Participant.user_id == target_user.id)
    )
    if existing_p.scalars().first():
        raise HTTPException(status_code=409, detail="User already a participant")

    participant = Participant(decision_id=did, user_id=target_user.id, role=dto.role)
    db.add(participant)
    
    audit = AuditLog(
        entity_id=did,
        actor_id=current_user.id,
        action="PARTICIPANT_INVITED",
        metadata_json={"target_user_id": str(target_user.id)}
    )
    db.add(audit)
    await db.commit()

    # Return updated list
    parts = await db.execute(select(Participant).filter(Participant.decision_id == did))
    return parts.scalars().all()

async def remove_participant(db: AsyncSession, decision_id: str, user_id: str, current_user: User):
    try:
        did = uuid.UUID(decision_id)
        uid = uuid.UUID(user_id)
    except ValueError:
        raise HTTPException(status_code=404, detail="Invalid UUID")

    if uid == current_user.id:
        raise HTTPException(status_code=400, detail="Cannot remove self")

    result = await db.execute(select(Decision).filter(Decision.id == did))
    if not result.scalars().first():
        raise HTTPException(status_code=404, detail="Decision not found")

    owner_check = await db.execute(
        select(Participant).filter(Participant.decision_id == did, Participant.user_id == current_user.id, Participant.role == "owner")
    )
    if not owner_check.scalars().first():
        raise HTTPException(status_code=403, detail="Only owner can remove")

    target_p = await db.execute(
        select(Participant).filter(Participant.decision_id == did, Participant.user_id == uid)
    )
    p_obj = target_p.scalars().first()
    if not p_obj:
        raise HTTPException(status_code=404, detail="Target not a participant")

    await db.delete(p_obj)
    
    audit = AuditLog(
        entity_id=did,
        actor_id=current_user.id,
        action="PARTICIPANT_REMOVED",
        metadata_json={"target_user_id": str(uid)}
    )
    db.add(audit)
    await db.commit()

    parts = await db.execute(select(Participant).filter(Participant.decision_id == did))
    return parts.scalars().all()
